File: apps/mobile/app/src/main/java/com/japaneselearning/mobile/data/remote/StudyApi.kt
package com.japaneselearning.mobile.data.remote

import retrofit2.http.Body
import retrofit2.http.DELETE
import retrofit2.http.GET
import retrofit2.http.POST
import retrofit2.http.PUT
import retrofit2.http.Path
import retrofit2.http.Query

interface StudyApi {
    @POST("auth/login")
    suspend fun login(@Body request: LoginRequest): LoginResponse

    @POST("auth/logout")
    suspend fun logout(): SuccessResponse

    @GET("auth/me")
    suspend fun me(): AuthMeResponse

    @GET("flashcard-sets")
    suspend fun listFlashcardSets(
        @Query("search") search: String? = null,
        @Query("page") page: Int = 1,
        @Query("limit") limit: Int = 20,
        @Query("favorite") favorite: Boolean? = null,
        @Query("tag") tag: String? = null,
    ): FlashcardSetsResponse

    @GET("flashcard-sets/{setId}")
    suspend fun getFlashcardSet(@Path("setId") setId: String): FlashcardSetDto

    @GET("review/summary")
    suspend fun getFlashcardReviewSummary(): FlashcardReviewSummaryDto

    @GET("review/queue")
    suspend fun getFlashcardReviewQueue(@Query("limit") limit: Int = 20): FlashcardReviewQueueResponseDto

    @POST("review/{cardId}")
    suspend fun submitFlashcardReview(
        @Path("cardId") cardId: String,
        @Body request: SubmitFlashcardReviewRequest,
    ): FlashcardReviewResponseDto

    @PUT("flashcard-sets/{setId}/favorite")
    suspend fun setFlashcardFavorite(
        @Path("setId") setId: String,
        @Body request: FavoriteRequest,
    ): FlashcardSetDto

    @PUT("flashcard-sets/{setId}/tags")
    suspend fun setFlashcardTags(
        @Path("setId") setId: String,
        @Body request: SetTagsRequest,
    ): FlashcardSetDto

    @GET("tags")
    suspend fun listTags(@Query("limit") limit: Int = 100): List<TagDto>

    @GET("exam-folders")
    suspend fun listExamFolders(): List<ExamFolderDto>

    @GET("exams")
    suspend fun listExams(
        @Query("folderId") folderId: String? = null,
        @Query("search") search: String? = null,
        @Query("page") page: Int = 1,
        @Query("limit") limit: Int = 20,
        @Query("favorite") favorite: Boolean? = null,
        @Query("tag") tag: String? = null,
    ): ExamsResponse

    @GET("exams/{examId}")
    suspend fun getExam(@Path("examId") examId: String): ExamDto

    @PUT("exams/{examId}/favorite")
    suspend fun setExamFavorite(
        @Path("examId") examId: String,
        @Body request: FavoriteRequest,
    ): ExamDto

    @PUT("exams/{examId}/tags")
    suspend fun setExamTags(
        @Path("examId") examId: String,
        @Body request: SetTagsRequest,
    ): ExamDto

    @POST("exams/{examId}/attempts")
    suspend fun startAttempt(@Path("examId") examId: String): LiveAttemptDto

    @GET("attempts/{attemptId}")
    suspend fun getAttempt(@Path("attemptId") attemptId: String): LiveAttemptDto

    @PUT("attempts/{attemptId}/answers")
    suspend fun saveAnswers(
        @Path("attemptId") attemptId: String,
        @Body request: SaveAnswersRequest,
    ): SuccessResponse

    @POST("attempts/{attemptId}/submit")
    suspend fun submitAttempt(
        @Path("attemptId") attemptId: String,
        @Body request: SubmitAttemptRequest = SubmitAttemptRequest(),
    ): ExamResultDto

    @GET("exam-review/mistakes")
    suspend fun getWrongAnswerReviewQueue(@Query("limit") limit: Int = 20): WrongAnswerReviewQueueDto

    @DELETE("exam-review/mistakes/{mistakeId}")
    suspend fun dismissWrongAnswer(
        @Path("mistakeId") mistakeId: String,
    ): SuccessResponse

    @DELETE("exam-review/mistakes")
    suspend fun clearWrongAnswers(@Query("examId") examId: String? = null): ClearMistakesResponse

    @POST("exam-review/practice")
    suspend fun startMistakePractice(@Body request: StartMistakePracticeRequest): LiveAttemptDto

    @GET("search")
    suspend fun search(
        @Query("q") query: String,
        @Query("limit") limit: Int = 20,
    ): SearchResultsDto

    @GET("dashboard/summary")
    suspend fun dashboard(): DashboardDto
}

File: apps/web/src/lib/study-api.ts
import {
  DashboardSummaryDto,
  ExamDto,
  ExamListItemDto,
  ExamFolderDto,
  ExamAttemptResultDto,
  FlashcardSetDto,
  FlashcardSetListItemDto,
  PaginatedResultDto,
  SearchResultsDto,
  LiveExamAttemptDto,
  RecentLearningResponseDto,
  TagDto,
  FlashcardReviewQueueResponseDto,
  FlashcardReviewResponseDto,
  FlashcardReviewSummaryDto,
  SubmitFlashcardReviewDto,
  StartMistakePracticeDto,
  WrongAnswerReviewQueueDto,
  CreateFlashcardDto,
  FlashcardDto,
  CreateDictionaryFavoriteDto,
  DictionaryFavoriteDto,
  DictionaryFavoriteListResponseDto,
  DictionaryLookupDirection,
  DictionaryLookupHistoryResponseDto,
  DictionaryLookupResponseDto,
  DictionarySuggestionResponseDto,
} from '@japanese-learning/contracts';
import { apiClient } from './api-client';
import { ExamListQuery, FlashcardListQuery } from './query-keys';

function buildQuery(params: Record<string, string | number | boolean | null | undefined>): string {
  const searchParams = new URLSearchParams();
  for (const [key, value] of Object.entries(params)) {
    if (value !== undefined && value !== null && value !== '') {
      searchParams.set(key, String(value));
    }
  }
  const encoded = searchParams.toString();
  return encoded ? `?${encoded}` : '';
}

const FORBIDDEN_LIVE_ATTEMPT_KEYS = new Set([
  'isCorrect',
  'correctOptionId',
  'answerKey',
  'correctAnswer',
]);

export function assertLiveAttemptPayload(payload: LiveExamAttemptDto): LiveExamAttemptDto {
  const pending: unknown[] = [payload];

  while (pending.length > 0) {
    const value = pending.pop();
    if (!value || typeof value !== 'object') continue;

    for (const [key, child] of Object.entries(value)) {
      if (FORBIDDEN_LIVE_ATTEMPT_KEYS.has(key)) {
        throw new Error('Live attempt response contains forbidden correctness metadata');
      }
      if (child && typeof child === 'object') pending.push(child);
    }
  }

  return payload;
}

export const studyApi = {
  dashboard: (signal?: AbortSignal) =>
    apiClient<DashboardSummaryDto>('/dashboard/summary', { signal }),

  recentLearning: (limit = 10, signal?: AbortSignal) =>
    apiClient<RecentLearningResponseDto>(`/recent-learning?limit=${limit}`, { signal }),

  flashcardSets: (query: FlashcardListQuery = {}, signal?: AbortSignal) =>
    apiClient<PaginatedResultDto<FlashcardSetListItemDto>>(
      `/flashcard-sets${buildQuery({
        search: query.search,
        page: query.page ?? 1,
        limit: query.pageSize ?? 20,
        sort: query.sort,
        favorite: query.favorite,
        tag: query.tag,
      })}`,
      { signal },
    ),

  flashcardSet: (setId: string, signal?: AbortSignal) =>
    apiClient<FlashcardSetDto>(`/flashcard-sets/${setId}`, { signal }),

  tags: (limit = 100, signal?: AbortSignal) =>
    apiClient<TagDto[]>(`/tags${buildQuery({ limit })}`, { signal }),

  reviewSummary: (signal?: AbortSignal) =>
    apiClient<FlashcardReviewSummaryDto>('/review/summary', { signal }),

  reviewQueue: (limit = 20, signal?: AbortSignal) =>
    apiClient<FlashcardReviewQueueResponseDto>(`/review/queue${buildQuery({ limit })}`, { signal }),

  submitReview: (cardId: string, body: SubmitFlashcardReviewDto, signal?: AbortSignal) =>
    apiClient<FlashcardReviewResponseDto>(`/review/${cardId}`, {
      method: 'POST',
      body: JSON.stringify(body),
      signal,
    }),

  mistakes: (limit = 20, examId?: string, signal?: AbortSignal) =>
    apiClient<WrongAnswerReviewQueueDto>(`/exam-review/mistakes${buildQuery({ limit, examId })}`, {
      signal,
    }),

  dismissMistake: (mistakeId: string, signal?: AbortSignal) =>
    apiClient<{ success: boolean }>(`/exam-review/mistakes/${mistakeId}`, {
      method: 'DELETE',
      signal,
    }),

  clearMistakes: (examId?: string, signal?: AbortSignal) =>
    apiClient<{ success: boolean; removedCount: number }>(
      `/exam-review/mistakes${buildQuery({ examId })}`,
      { method: 'DELETE', signal },
    ),

  startMistakePractice: (body: StartMistakePracticeDto, signal?: AbortSignal) =>
    apiClient<LiveExamAttemptDto>('/exam-review/practice', {
      method: 'POST',
      body: JSON.stringify(body),
      signal,
    }).then(assertLiveAttemptPayload),

  setFlashcardTags: (setId: string, tags: string[], signal?: AbortSignal) =>
    apiClient<FlashcardSetDto>(`/flashcard-sets/${setId}/tags`, {
      method: 'PUT',
      body: JSON.stringify({ tags }),
      signal,
    }),

  setFlashcardFavorite: (setId: string, favorite: boolean, signal?: AbortSignal) =>
    apiClient<FlashcardSetDto>(`/flashcard-sets/${setId}/favorite`, {
      method: 'PUT',
      body: JSON.stringify({ favorite }),
      signal,
    }),

  examFolders: (signal?: AbortSignal) => apiClient<ExamFolderDto[]>('/exam-folders', { signal }),

  exams: (query: ExamListQuery = {}, signal?: AbortSignal) =>
    apiClient<PaginatedResultDto<ExamListItemDto>>(
      `/exams${buildQuery({
        folderId: query.folderId,
        search: query.search,
        page: query.page ?? 1,
        limit: query.pageSize ?? 20,
        sort: query.sort,
        favorite: query.favorite,
        tag: query.tag,
      })}`,
      { signal },
    ),

  exam: (examId: string, signal?: AbortSignal) =>
    apiClient<ExamDto>(`/exams/${examId}`, { signal }),

  setExamTags: (examId: string, tags: string[], signal?: AbortSignal) =>
    apiClient<ExamDto>(`/exams/${examId}/tags`, {
      method: 'PUT',
      body: JSON.stringify({ tags }),
      signal,
    }),

  setExamFavorite: (examId: string, favorite: boolean, signal?: AbortSignal) =>
    apiClient<ExamDto>(`/exams/${examId}/favorite`, {
      method: 'PUT',
      body: JSON.stringify({ favorite }),
      signal,
    }),

  search: (query: string, limit = 30, signal?: AbortSignal) =>
    apiClient<SearchResultsDto>(`/search${buildQuery({ q: query.trim(), limit })}`, { signal }),

  dictionaryLookup: (
    query: string,
    direction: DictionaryLookupDirection = DictionaryLookupDirection.AUTO,
    limit = 20,
    includeExamples = false,
    signal?: AbortSignal,
  ) =>
    apiClient<DictionaryLookupResponseDto>(
      `/lookup${buildQuery({ q: query.trim(), direction, limit, includeExamples })}`,
      { signal },
    ),

  dictionarySuggestions: (
    query: string,
    direction: DictionaryLookupDirection = DictionaryLookupDirection.AUTO,
    limit = 10,
    signal?: AbortSignal,
  ) =>
    apiClient<DictionarySuggestionResponseDto>(
      `/lookup/suggest${buildQuery({ q: query.trim(), direction, limit })}`,
      { signal },
    ),

  dictionaryHistory: (limit = 10, signal?: AbortSignal) =>
    apiClient<DictionaryLookupHistoryResponseDto>(`/lookup/history${buildQuery({ limit })}`, {
      signal,
    }),

  clearDictionaryHistory: (signal?: AbortSignal) =>
    apiClient<{ deleted: number }>('/lookup/history', { method: 'DELETE', signal }),

  dictionaryFavorites: (limit = 20, offset = 0, signal?: AbortSignal) =>
    apiClient<DictionaryFavoriteListResponseDto>(
      `/lookup/favorites${buildQuery({ limit, offset })}`,
      { signal },
    ),

  saveDictionaryFavorite: (body: CreateDictionaryFavoriteDto, signal?: AbortSignal) =>
    apiClient<DictionaryFavoriteDto>('/lookup/favorites', {
      method: 'POST',
      body: JSON.stringify(body),
      signal,
    }),

  removeDictionaryFavorite: (favoriteId: string, signal?: AbortSignal) =>
    apiClient<{ success: true; id: string }>(`/lookup/favorites/${favoriteId}`, {
      method: 'DELETE',
      signal,
    }),

  createFlashcard: (setId: string, body: CreateFlashcardDto, signal?: AbortSignal) =>
    apiClient<FlashcardDto>(`/flashcard-sets/${setId}/cards`, {
      method: 'POST',
      body: JSON.stringify(body),
      signal,
    }),

  liveAttempt: (attemptId: string, signal?: AbortSignal) =>
    apiClient<LiveExamAttemptDto>(`/attempts/${attemptId}`, { signal }).then(
      assertLiveAttemptPayload,
    ),

  startAttempt: (examId: string, signal?: AbortSignal) =>
    apiClient<LiveExamAttemptDto>(`/exams/${examId}/attempts`, {
      method: 'POST',
      signal,
    }).then(assertLiveAttemptPayload),

  submitAttempt: (attemptId: string, body: unknown, signal?: AbortSignal) =>
    apiClient<ExamAttemptResultDto>(`/attempts/${attemptId}/submit`, {
      method: 'POST',
      body: JSON.stringify(body),
      signal,
    }),
};
